File: ruby/tut6.rb
#!/usr/bin/env ruby

# XMMS2 - X Music Multiplexer System
# Copyright (C) 2003-2006 XMMS2 Team

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.

# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.

# This file is a part of the XMMS2 client tutorial #6

# This is the async version of tutorial 2. You will need Ruby-GLib2 to run this
# tutorial. (However, GLib is by no means the only event loop you can use for an
# asynchronous client.)
require 'glib2'
require 'xmmsclient'
# XMMS2 comes with a few mainloop integration tools we can use. Even better is
# that they are bound in ruby. Here is the one for GLib2.
require 'xmmsclient_glib'

xmms = Xmms::Client.new('tutorial6')

begin
	xmms.connect(ENV['XMMS_PATH'])
rescue Xmms::Client::ClientError
	puts 'Failed to connect to XMMS2 daemon.'
	puts 'Please make sure xmms2d is running and using the correct IPC path.'
	exit
end

# Here we can initialize a GLib::MainLoop instance for later use. Ignore the
# arguments passed, they're unimportant boilerplate stuff.
ml = GLib::MainLoop.new(nil, false)

# GLib mainloop integration is a breeze with the tools provided for GLib in
# XMMS2. Just one function call sets it all up.
xmms.add_to_glib_mainloop

# Just like in the second tutorial, the Xmms::Result object is stored in res.
res = xmms.playback_current_id

# The difference in an asynchronous (async) client is that the wait method is
# not used at all and other code can be run while the daemon and client
# communicate.
# So, when we want our finished result, a callback is called. That means we need
# to set up a callback Proc, which we do with the notifier method. Here is where
# any Rubyist will start to feel the love: Ruby blocks!
res.notifier do |res|
	id = res.value

	if(id == 0)
		puts 'There is no current ID. XMMS2 is probably not playing anything.'
	else
		puts "Currently playing ID: #{id}"
	end

	# The ID should only appear once. The mainloop should die here.
	ml.quit
end

puts <<END
Here's something else that's pretty interesting. Besides the initial connection
and instantiation of the Xmms::Client object, nothing happens inside an async
client until the mainloop is invoked. It's the mainloop's job to talk to the
daemon and take in any data that needs to come in or send out any data that is
waiting to be sent.

You will notice that this message is printed before the ID.

END

# Start the GLib mainloop and the ID should be printed.
ml.run

File: c++/tut6.cc
/*  XMMS2 - X Music Multiplexer System
 *  Copyright (C) 2003-2006 XMMS2 Team
 * 
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *                   
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  This file is a part of the XMMS2 client tutorial #6
 *  Introducing asynchronous clients
 */

#include <xmmsclient/xmmsclient++.h>
#include <cstdlib>
#include <iostream>

// Our little class to handle everything in a nice object-oriented way.
class MyClient
{

	public:

		MyClient();
		~MyClient();

		/*
		 * Here's our callback function declarations.
		 * Note that a callback function must return a bool.
		 * The meaning of this return value is explained later.
		 */
		bool my_current_id( const int& id );
		bool error_handler( const std::string& error );

	private:
		Xmms::Client client_;

};

MyClient::MyClient() : client_( "tutorial6" )
{

	/*
	 * In an asynchronous client we still connect normally
	 * Read up on this in earlier tutorials if you need.
	 */
	client_.connect( std::getenv( "XMMS_PATH" ) );

	/*
	 * The big difference between a sync and an async client is that the
	 * async client works with callbacks. When you send a command you need
	 * to set a callback for it.
	 *
	 * Using asynchronous functions works almost the same way as the
	 * synchronous versions - in fact the function used is the same.
	 * The difference is the way you use the return value of a function.
	 * All clientlib functions actually return an adapter class which
	 * implicitly converts into a value in synchronous mode. This adapter
	 * class is also used to bind callback functions which are then called
	 * later with the result from the function you originally called.
	 *
	 * Callback functions can be set iwth operator() of the adapter class
	 * (which also finishes the call, you can't add more callbacks afterwards)
	 * or with connect() and connectError() functions. If you use connect*
	 * functions or don't want to bind any callback functions, just use
	 * operator() with no arguments to finish the call,
	 *
	 * You can set normal functions and member functions as callback,
	 * but there is a special syntax for member function pointers.
	 * In order to *bind* a member function to a callback, you need to use
	 * Xmms::bind which takes a function pointer and a pointer to the
	 * object from which the function is to be called.
	 *
	 * Normal functions and static member functions can be bound with
	 * just &my_function, like any other function pointer.
	 */
	client_.playback.currentID()( Xmms::bind( &MyClient::my_current_id, this ),
	                              Xmms::bind( &MyClient::error_handler, this ) );

	/*
	 * xmmsclient++ has its own mainloop which you can use if you don't
	 * want to depend on any other library, like gtkmm (glib), qt, ecore
	 * and corefoundation.
	 * Just invoke this and your client will enter the mainloop and the
	 * signals/broadcasts and other functions will start rolling.
	 */
	client_.getMainLoop().run();

}

MyClient::~MyClient() { }

bool MyClient::my_current_id( const int& id )
{

	/*
	 * This is our callback function which will eventually get an answer
	 * for the question we set earlier.
	 */
	std::cout << "Current id is " << id << std::endl;

	/* 
	 * There are three kinds of asynchronous functions.
	 *  - Normal functions
	 *  - Signals
	 *  - Broadcasts
	 *
	 * signals and broadcasts have the function type as the first
	 * word on the function, such as Client::broadcastQuit.
	 *
	 * Depending on which type of function we're handling with
	 * the meaning of this return value varies a little.
	 *  - Normal function : return value doesn't have any specific meaning.
	 *  - Signal
	 *    - true  : signal will be reset and you'll get a new answer soon.
	 *    - false : signal will be eliminated.
	 *  - Broadcast
	 *    - true  : broadcast will continue.
	 *    - false : broadcast is disconnected. You won't be getting any more
	 *              of these.
	 */
	return false;

}

bool MyClient::error_handler( const std::string& error )
{

	/*
	 * This is the error callback function which will get called
	 * if there was an error in the process.
	 */
	std::cout << "Error: " << error << std::endl;
	return false;

}

int
main()
{

	try {
		MyClient myclient;
	}
	catch( Xmms::connection_error& err ) {
		std::cout << "Connection failed: " << err.what() << std::endl;
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;

}
